File: 4.1-calculos-matematicos/Prueba-1/Variacion-2/python/main.py
import time
import js  # type: ignore
import tracemalloc
from pyscript import display  # type: ignore
import numpy as np  # Importamos numpy


def create_matrix(size):
    return np.random.rand(size, size)


def multiply_matrices(size):
    tracemalloc.start()
    A = create_matrix(size)
    B = create_matrix(size)
    C = np.zeros((size, size))

    start = time.time()

    C = np.dot(A, B)

    end = time.time()

    # ET (Execution Time)
    execution_time = (end - start) * 1000
    display(f"ET: {round(execution_time, 2)} ms", target="pyscript-output")

    # PLT
    js.endTimerWebAssembly()

    # RAM
    memory_usage = tracemalloc.get_traced_memory()[1] / (1024 * 1024)
    display(f"RAM: {round(memory_usage, 2)} MB",
            target="pyscript-output")
    tracemalloc.stop()


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    multiply_matrices(500)
    multiply_matrices(1000)
    multiply_matrices(2000)

File: 4.1-calculos-matematicos/Prueba-1/Variacion-2/javascript/main.js
function runNodeBenchmark() {
    clearCell("nodeJs-output");
    fetch("http://localhost:3000/")
        .then(response => response.json())
        .then(data => {
            let outputDiv = document.getElementById("nodeJs-output");

            let timeDiv = document.createElement("div");
            timeDiv.textContent = `ET: ${data.executionTime} ms`;
            outputDiv.appendChild(timeDiv);

            let cpuDiv = document.createElement("div");
            cpuDiv.textContent = `CPU: ${data.cpuUsage} %`;
            outputDiv.appendChild(cpuDiv);

            let memoryDiv = document.createElement("div");
            memoryDiv.textContent = `RAM: ${data.memoryUsage} MB`;
            outputDiv.appendChild(memoryDiv);

        })
        .catch(error => console.error("Error:", error));
}

async function multiplyMatrices(size) {

    let A = tf.randomNormal([size, size]);
    let B = tf.randomNormal([size, size]);

    let start = performance.now();

    let C = tf.matMul(A, B);

    await C.data();

    let end = performance.now();

    // ET (Execution Time)
    let resultTime = Number((end - start).toFixed(2));
    let result = `ET: ${resultTime} ms`;
    let resultDiv = document.createElement("div");
    resultDiv.textContent = result;
    document.getElementById("javascript-output").appendChild(resultDiv);
}

async function runJSBenchmark() {
    clearCell("javascript-output");
    document.getElementById("javascript-output").textContent = "";
    await multiplyMatrices(1000);

    let memoryDiv = document.createElement("div");
    memoryDiv.textContent = getMemoryUsageJS();
    document.getElementById("javascript-output").appendChild(memoryDiv);
}

function getMemoryUsageJS() {
    // RAM
    if (performance.memory) {
        let memoryUsed = performance.memory.usedJSHeapSize / (1024 * 1024);
        return `RAM: ${memoryUsed.toFixed(2)} MB`;
    } else {
        return `RAM: Unsupported measurement in this browser.`;
    }
}

function clearCell(elementId) {
    document.getElementById(elementId).textContent = "";
}